meter/e3dc.go

```diff
@@ -0,0 +1,129 @@
+package meter
+
+import (
+	"time"
+
+	"github.com/evcc-io/evcc/api"
+	"github.com/evcc-io/evcc/util"
+	"github.com/evcc-io/evcc/util/request"
+	"github.com/evcc-io/evcc/util/templates"
+	"github.com/spali/go-rscp/rscp"
+	"github.com/spf13/cast"
+)
+
+type E3dc struct {
+	// capacity float64
+	usage templates.Usage // TODO check if we really want to depend on templates
+	conn  *rscp.Client
+}
+
+func init() {
+	registry.Add("e3dc-2", NewE3dcFromConfig)
+}
+
+//go:generate go run ../cmd/tools/decorate.go -f decorateE3dc -b *E3dc -r api.Meter -t "api.Battery,Soc,func() (float64, error)"
+
+func NewE3dcFromConfig(other map[string]interface{}) (api.Meter, error) {
+	cc := struct {
+		capacity `mapstructure:",squash"`
+		Usage    templates.Usage
+		Address  string
+		Port     uint16
+		Username string
+		Password string
+		Key      string
+		Timeout  time.Duration
+	}{
+		Timeout: request.Timeout,
+	}
+
+	if err := util.DecodeOther(other, &cc); err != nil {
+		return nil, err
+	}
+
+	cfg := rscp.ClientConfig{
+		Address:           cc.Address,
+		Port:              cc.Port,
+		Username:          cc.Username,
+		Password:          cc.Password,
+		Key:               cc.Key,
+		ConnectionTimeout: cc.Timeout,
+		SendTimeout:       cc.Timeout,
+		ReceiveTimeout:    cc.Timeout,
+	}
+
+	return NewE3dc(cc.Usage, cfg)
+}
+
+func NewE3dc(usage templates.Usage, cfg rscp.ClientConfig) (api.Meter, error) {
+	// util.NewLogger("e3dc")
+	conn, err := rscp.NewClient(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	res := &E3dc{
+		usage: usage,
+		conn:  conn,
+	}
+
+	// decorate api.BatterySoc
+	var batterySoc func() (float64, error)
+	// var batteryCapacity func() float64
+	if usage == templates.UsageBattery {
+		batterySoc = res.batterySoc
+		// batteryCapacity = res.batteryCapacity
+
+		// const TAG_BAT_REQ_SPECIFICATION = rscp.Tag(0x03000043)
+		// msg, err := res.conn.Send(*rscp.NewMessage(TAG_BAT_REQ_SPECIFICATION, nil))
+		// if err != nil {
+		// 	return nil, err
+		// }
+
+		// cap, err := cast.ToFloat64E(msg.Value)
+		// if err != nil {
+		// 	return nil, err
+		// }
+
+		// res.capacity = cap / 1e3
+	}
+
+	return decorateE3dc(res, batterySoc), nil
+}
+
+func (m *E3dc) CurrentPower() (float64, error) {
+	var tag rscp.Tag
+	sign := 1.0
+
+	switch m.usage {
+	case templates.UsageGrid:
+		tag = rscp.EMS_REQ_POWER_GRID
+	case templates.UsagePV:
```

Comment:
For PV, it would also make sense to add the tag EMS_REQ_POWER_ADD. This is particularly important if additional e.g. external inverters are connected to an E3DC. These are otherwise not included in the calculation. Comparable to https://github.com/evcc-io/evcc/blob/72724bc5dd19ac8d5ea9b40c35c04303a9dfee84/templates/definition/meter/e3dc.yaml#L39 in the previous implementation.

https://github.com/spali/go-rscp/blob/d8133bf98dd3dab93accb10370def10fb8a77172/rscp/tag.go#L42

Reply:
Bitte PR. Keine Ahnung wie man mehrere Tags in einem Rutsch abfragen kann.

Reply:
Leibt die Reihenfolge in der Antwort erhalten oder wie finde ich das Mapping? Liefert ein (Container) Request dann mehrere Ergebnisse?

Mir ist das völlig unklar. Ein Codebeispiel würde Klarheit schaffen.